### PR TITLE
Create Ntasks Flexibility Function

### PR DESCRIPTION
**Pull request by manishvenu**
Create a function that adjusts the number of ntasks based on grid size.

Companion PR that has the demo that resulted in this function (in CrocoDash): https://github.com/CROCODILE-CESM/CrocoDash/pull/28

Questions

- Should I call it "ntasks" or the function should called set_cores_based_on_grid instead
- Currently, cvars has OCN_NX & OCN_NY, I have only set NTASKS_OCN based on that, but should it be for every component?
- In Regional CaseGen, the idea was to request NTASKS as a multiple of the number of cores per node (128) because of some efficiency reason, currently it's set to request more NTASKS in multiples  of 16 if there is more points per core than recommended , but should that actually be 128? The reason it is 16  is because, for reducing the number of NTASKS, 16 works well b/c we need a denomination lower than the cores per node to stop the "MPP_DEFINE_DOMAINS(mpp_compute_extent): domain extents must be positive definite." error.

### CODE REVIEW

**Comment by manishvenu** [comment]
Notes on buildnml MOM_interface check for this error: 

- Error exists in the FMS function mpp_compute_extent
- It's called from MOM6 only in MOM_domains in determine_land_blocks, which is called from gen_auto_mask_table
- gen_auto_mask_table iterates on determine_land_blocks by changing the equivalent parameter to NTASKS_OCN. It divides NTASKS_OCN by ocean_frac and iterates from their down to NTASKS_OCN. There's check to skip a few iterations nd a check to stop the for loop. Each iteration calls determine_land_blocks, which calls compute_extent. Any of these iterations could throw the error.

**Comment by alperaltuntas** [comment]
I wonder whether `num_points` should be a function argument for generality purposes. And we may as well rename this function as `_set_ntasks_based_on_grid`. 

I also wonder whether we should take `nk` into account.

Also, I believe this may be a good time to introduce property-based testing. There is a divide-by-zero bug that `hypothesis` found quickly and easily. With some property assertions, it may find more issues. Let's start using it!

> Should I call it "ntasks" or the function should called set_cores_based_on_grid instead

Let's use `ntasks` to be consistent with CESM variable naming.

> Currently, cvars has OCN_NX & OCN_NY, I have only set NTASKS_OCN based on that, but should it be for every component?

Yes, and I think having `num_points` as the function argument should facilitate that genericity.

> In Regional CaseGen, the idea was to request NTASKS as a multiple of the number of cores per node (128) because of some efficiency reason, currently it's set to request more NTASKS in multiples of 16 if there is more points per core than recommended , but should that actually be 128? The reason it is 16 is because, for reducing the number of NTASKS, 16 works well b/c we need a denomination lower than the cores per node to stop the "MPP_DEFINE_DOMAINS(mpp_compute_extent): domain extents must be positive definite." error.

I think, initially, the idea of "requesting NTASKS as a multiple of the number of cores per node" stemmed from the fact that we get charged for the entire node even if we only use a portion of it. But, yeah, if that results in MPP_DEFINE_DOMAINS error, then less incrementations make sense. Either way, I think we need property based testing to catch these and similar edge cases.

File: visualCaseGen/custom_widget_types/case_creator.py [hunk 1]
```diff
@@ -523,7 +523,7 @@ def _set_ntasks_ocean_based_on_grid(self, do_exec, min_points_per_core = 32, max
             cores = cores - iteration_amount
             pts_per_core = num_points/cores
 
-        xmlchange("NTASKS_OCN",cores, do_exec, self._is_non_local(), self._out)
+        xmlchange("NTASKS",cores, do_exec, self._is_non_local(), self._out)
```
Comment (alperaltuntas) [inline visualCaseGen/custom_widget_types/case_creator.py:526]: I think this function should only return `cores`, and should not call xmlchange. It should be up to callee to use the returned `cores` in an `xmlchange` call for NTASKS_OCN or whichever NTASKS_??? is appropriate (again, for generality purposes.) With this suggestion, this function can become a staticmethod since we wouldn't need `self` anymore (aside from logging, which can also happen outside this function). This would make this function compact and easy to unit test.

**Comment by manishvenu** [comment]
@alperaltuntas How do these updates look?

File: visualCaseGen/custom_widget_types/case_creator.py [hunk 2]
```diff
@@ -502,6 +502,33 @@ def _apply_all_xmlchanges(self, do_exec):
...
+            print(f"{COMMENT}Apply NTASK grid xml changes:{RESET}\n")
+            xmlchange("NTASKS_OCN",cores, do_exec, self._is_non_local(), self._out)
+
+    @staticmethod
+    def _calc_cores_based_on_grid( num_points, min_points_per_core = 32, max_points_per_core = 800):
+        """Calculate the number of cores based on the grid size."""
+
+
+        cores = 128 # Start from 128 which is the default 128 cores per node in derecho
+        iteration_amount = 16
```
Comment (alperaltuntas) [inline visualCaseGen/custom_widget_types/case_creator.py:518]: As I understand it, min cores of 16 in test_cores_case_creator.py
 is dictated by this hard coded iteration_amount, right? Min cores of 16 is too restrictive, I think. Can we set iteration amount based on `num_points`, e.g., `iteration_amount = num_points / min_points_per_core`. Also, I think we need to add hypothesis into the conda environment.

**Review by alperaltuntas — CHANGES_REQUESTED** [review]

File: visualCaseGen/custom_widget_types/case_creator.py (hunk 2 above)
Comment (manishvenu) [reply visualCaseGen/custom_widget_types/case_creator.py:518]: Yup, shoot I'll do this

**Review by alperaltuntas — APPROVED** [review]
